### PR TITLE
feature/SOF-7427 feat: ASE distance norm calculator

### PR DESCRIPTION
**Pull request by VsevolodX**
- feat: add ASE calculator 
- add docstrings explaining  
- add required forces defintions and calculations

### CODE REVIEW

File: src/py/mat3ra/made/tools/calculate.py [hunk 1]
```diff
@@ -167,3 +167,100 @@ def calculate_norm_of_distances(material: Material, shadowing_radius: float = 2.
     substrate_coordinates_values = np.array(substrate_atoms_surface_coordinates.values)
 
     return get_norm_of_distances_between_coordinates(film_coordinates_values, substrate_coordinates_values)
+
+
+class SurfaceDistanceCalculator(ASECalculator):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:172]: FixedZ

File: src/py/mat3ra/made/tools/calculate.py [hunk 2]
```diff
@@ -167,3 +167,100 @@ def calculate_norm_of_distances(material: Material, shadowing_radius: float = 2.
...
+    def calculate(self, atoms: Optional[ASEAtoms] = None, properties=["energy"], system_changes=ase_all_changes):
+        if atoms is None:
+            atoms = self.atoms.copy()
+
+        atoms = self._add_constraints(atoms)
+        constraints = atoms.constraints
+
+        ASECalculator.calculate(self, atoms, properties, system_changes)
+        material = Material(from_ase(atoms))
+        norm = calculate_norm_of_distances(material, self.shadowing_radius)
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:252]: energies

File: src/py/mat3ra/made/tools/calculate.py [hunk 3]
```diff
@@ -167,3 +167,100 @@ def calculate_norm_of_distances(material: Material, shadowing_radius: float = 2.
...
+            substrate_indices = [i for i, tag in enumerate(atoms.get_tags()) if tag == 0]
+            constraints.append(ASEFixAtoms(indices=substrate_indices))
+        if self.fix_z:
+            all_indices = list(range(len(atoms)))
+            constraints.append(ASEFixedPlane(indices=all_indices, direction=[0, 0, 1]))
+
+        atoms.set_constraint(constraints)
+        return atoms
+
+    def _calculate_forces(self, atoms: ASEAtoms, norm: float) -> np.ndarray:
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:228]: Maybe we can remove this

File: src/py/mat3ra/made/tools/calculate.py (hunk 1 above)
Comment (VsevolodX) [reply src/py/mat3ra/made/tools/calculate.py:172]: Uhm, we pass this as a paramter: whether to fix_z, fix_substrate or not

File: src/py/mat3ra/made/tools/calculate.py [hunk 4]
```diff
@@ -173,3 +173,107 @@ def calculate_film_substrate_interaction_metric(
...
+
+class SurfaceDistanceCalculator(ASECalculator):
+    """
+    ASE calculator that computes the norm of distances between interfacial gap facing atoms
+    of the film and the substrate.
+
+    Args:
+        shadowing_radius (float): The radius for atom to shadow underlying from being considered surface, in Angstroms.
+        force_constant (float): The force constant for the finite difference approximation of the forces.
+        fix_substrate (bool): Whether to fix the substrate atoms.
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:186]: is_substrate_fixed, is_z_axis_fixed

File: src/py/mat3ra/made/tools/calculate.py [hunk 5]
```diff
@@ -173,3 +173,107 @@ def calculate_film_substrate_interaction_metric(
...
+        fix_z (bool): Whether to fix atoms movement in the z direction.
+        symprec (float): The symmetry precision for the ASE calculator.
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:188]: What is that - let's put a link to the original explanation

File: src/py/mat3ra/made/tools/calculate.py [hunk 6]
```diff
@@ -173,3 +173,107 @@ def calculate_film_substrate_interaction_metric(
...
+    def calculate(self, atoms: Optional[ASEAtoms] = None, properties=["energy"], system_changes=ase_all_changes):
+        if atoms is None:
+            atoms = self.atoms.copy()
+
+        atoms = self._add_constraints(atoms)
+        constraints = atoms.constraints
+
+        ASECalculator.calculate(self, atoms, properties, system_changes)
+        material = Material(from_ase(atoms))
+        energy = calculate_film_substrate_interaction_metric(material, self.shadowing_radius)
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:265]: This should depend on the calculator

File: src/py/mat3ra/made/tools/calculate.py (hunk 1 above)
Comment (timurbazhirov) [reply src/py/mat3ra/made/tools/calculate.py:178]: FilmSubstrateDistanceCalculator or InterfacialSeparationDistanceCalculator

File: src/py/mat3ra/made/tools/calculate.py [hunk 7]
```diff
@@ -173,3 +173,107 @@ def calculate_film_substrate_interaction_metric(
...
+
+    atoms.calc = calc
+    opt = BFGS(atoms)
+    opt.run(fmax=0.05)
+    ```
+    Args:
+        shadowing_radius (float): Radius for atoms shadowing underlying from being treated as a surface, in Angstroms.
+        force_constant (float): The force constant for the finite difference approximation of the
+    Note:
+        Built following: https://wiki.fysik.dtu.dk/ase/development/calculators.html
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:204]: remove :

File: src/py/mat3ra/made/tools/calculate.py [hunk 8]
```diff
@@ -173,3 +173,107 @@ def calculate_film_substrate_interaction_metric(
...
+
+    @decorator_convert_material_args_kwargs_to_atoms
+    def calculate(self, atoms: Optional[ASEAtoms] = None, properties=["energy"], system_changes=ase_all_changes):
+        if atoms is None:
+            atoms = self.atoms.copy()
+
+        atoms = self._add_constraints(atoms)
+        constraints = atoms.constraints
+
+        ASECalculator.calculate(self, atoms, properties, system_changes)
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:263]: use super

File: src/py/mat3ra/made/tools/calculate.py [hunk 9]
```diff
@@ -173,3 +173,107 @@ def calculate_film_substrate_interaction_metric(
...
+                atoms_plus = atoms.copy()
+                atoms_plus.positions[i, j] += dx
+                material_plus = Material(from_ase(atoms_plus))
+                energy_plus = calculate_film_substrate_interaction_metric(material_plus, self.shadowing_radius)
+
+                forces[i, j] = -self.force_constant * (energy_plus - energy) / dx
+
+        return forces
+
+    @decorator_convert_material_args_kwargs_to_atoms
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:255]: No point of having this

File: src/py/mat3ra/made/tools/calculate.py [hunk 10]
```diff
@@ -173,3 +173,107 @@ def calculate_film_substrate_interaction_metric(
...
+
+        The calculate method is responsible for computing the energy and forces (if requested).
+        Forces are estimated using a finite difference method, which is a simple approximation
+        and might not be the most accurate or efficient for all cases.
+    """
+
+    implemented_properties = ["energy", "forces"]
+
+    def __init__(
+        self,
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:214]: should have material_calculator as parameter

File: src/py/mat3ra/made/tools/calculate.py [hunk 11]
```diff
@@ -173,3 +173,107 @@ def calculate_film_substrate_interaction_metric(
...
+
+
+class SurfaceDistanceCalculator(ASECalculator):
+    """
+    ASE calculator that computes the norm of distances between interfacial gap facing atoms
+    of the film and the substrate.
+
+    Args:
+        shadowing_radius (float): The radius for atom to shadow underlying from being considered surface, in Angstroms.
+        force_constant (float): The force constant for the finite difference approximation of the forces.
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:185]: Remove

File: src/py/mat3ra/made/tools/calculate/calculators.py [hunk 12]
```diff
@@ -120,3 +122,174 @@ def get_energy(
         substrate_coordinates_values = np.array(substrate_surface_atom_coordinates.values)
 
         return interaction_function(film_coordinates_values, substrate_coordinates_values)
+
+
+class FixFilmRigidXY(BaseModel):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate/calculators.py:127]: class InterfaceConstraint

class RigidFilmXYConstraint(InterfaceConstraint)

File: src/py/mat3ra/made/tools/calculate/calculators.py [hunk 13]
```diff
@@ -120,3 +123,112 @@ def get_energy(
...
+        self.shadowing_radius = shadowing_radius
+        self.fix_substrate = is_substrate_fixed
+        self.fix_z = is_z_axis_fixed
+        self.symprec = symprec
+        self.material_calculator = material_calculator
+
+    def _add_constraints(self, atoms: ASEAtoms) -> ASEAtoms:
+        constraints: List[Union[ASEFixAtoms, ASEFixedPlane, RigidFilmXYConstraint]] = []
+        if self.fix_substrate:
+            substrate_indices = [i for i, tag in enumerate(atoms.get_tags()) if tag == 0]
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate/calculators.py:184]: We should abstract this into `get_interface_part_indices`

File: src/py/mat3ra/made/tools/calculate/calculators.py [hunk 14]
```diff
@@ -120,3 +123,112 @@ def get_energy(
...
+            constraints.append(ASEFixAtoms(indices=substrate_indices))
+        if self.fix_z:
+            all_indices = list(range(len(atoms)))
+            constraints.append(ASEFixedPlane(indices=all_indices, direction=[0, 0, 1]))
+
+        film_indices = [i for i, tag in enumerate(atoms.get_tags()) if tag == 1]
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate/calculators.py:190]: Same as above

File: src/py/mat3ra/made/tools/calculate/calculators.py [hunk 15]
```diff
@@ -120,3 +123,112 @@ def get_energy(
         substrate_coordinates_values = np.array(substrate_surface_atom_coordinates.values)
 
         return interaction_function(film_coordinates_values, substrate_coordinates_values)
+
+
+class FilmSubstrateDistanceASECalculator(ASECalculator):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate/calculators.py:128]: Let's put in `ase` subfolder for this and constraints into it

File: src/py/mat3ra/made/tools/calculate/constraints.py [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        """
+        Adjust the forces on the atoms in the system.
+
+        Args:
+            forces (numpy.ndarray): The forces array to adjust in place.
+        """
+        raise NotImplementedError
+
+
+class RigidFilmXYConstraint(InterfaceConstraint):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate/constraints.py:37]: InterfaceConstraint

File: src/py/mat3ra/made/tools/calculate/calculators.py (hunk 13 above)
Comment (VsevolodX) [reply src/py/mat3ra/made/tools/calculate/calculators.py:184]: Ok, but locally. Since it's ASE atoms, not Material

**Review by timurbazhirov — APPROVED** [review]

**Pull request merged**